llama.cpp

```diff
@@ -0,0 +1,1369 @@
+#include <cstdio>
+#include <cstring>
+#include <fstream>
+#include <map>
+#include <string>
+#include <vector>
+#include <regex>
+#include <optional>
+
+// TODO: move somewhere else
+#define QK 32
+
+// Define std::make_unique if it's not available
+// (e.g. on C++11)
+#ifndef __cpp_lib_make_unique
+template<typename T, typename... Args>
+std::unique_ptr<T> make_unique(Args&&... args) {
+    return std::unique_ptr<T>(new T(std::forward<Args>(args)...));
+}
+#else
+using std::make_unique;
+#endif
+
+// determine number of model parts based on the dimension
+static const std::map<int, int> LLAMA_N_PARTS = {
+    { 4096, 1 },
+    { 5120, 2 },
+    { 6656, 4 },
+    { 8192, 8 },
+};
+
+struct llama_layer {
+    // normalization
+    struct ggml_tensor * attention_norm;
+
+    // attention
+    struct ggml_tensor * wq;
+    struct ggml_tensor * wk;
+    struct ggml_tensor * wv;
+    struct ggml_tensor * wo;
+
+    // normalization
+    struct ggml_tensor * ffn_norm;
+
+    // ff
+    struct ggml_tensor * w1;
+    struct ggml_tensor * w2;
+    struct ggml_tensor * w3;
+};
+struct llama_model {
+    llama_hparams hparams;
+
+    struct ggml_tensor * tok_embeddings;
+
+    struct ggml_tensor * norm;
+    struct ggml_tensor * output;
+
+    std::vector<llama_layer> layers;
+
+    // key + value memory
+    struct ggml_tensor * memory_k;
+    struct ggml_tensor * memory_v;
+
+    //
+    struct ggml_context * ctx;
+    std::map<std::string, struct ggml_tensor *> tensors;
+};
+struct llama_state
+{
+    // Timers
+    struct timing {
+        int64_t t_load_us = 0;
+
+        int64_t t_sample_us = 0;
+        int64_t t_predict_us = 0;
+    } timing;
+
+    // Random number generator
+    std::mt19937 rng{};
+
+    // Tokens
+    std::vector<gpt_vocab::id> embd{};
+    std::vector<gpt_vocab::id> embd_inp{};
+    std::vector<gpt_vocab::id> last_n_tokens{};
+
+    // Logits from inference
+    std::vector<float> logits{};
+
+    // Counters
+    int input_consumed = 0;
+    int remaining_tokens = 0;
+    int n_past = 0;
+    size_t mem_per_token = 0;
+
+    // Flag set after initialization
+    bool is_initialized = false;
+};
+struct llama_context
+{
+    ggml_type wtype = ggml_type::GGML_TYPE_F16; // weight type (FP32 or FP16)
+
+    llama_model model{};
+    gpt_vocab vocab{};
+    gpt_params params{};
+
+    std::unique_ptr<llama_state> state = nullptr;
+
+    // default constructor
+    llama_context() = default;
+    // constructor
+    llama_context(llama_model&& model, gpt_vocab&& vocab, const gpt_params& params):
+        wtype(ggml_type::GGML_TYPE_F16),
+        model(std::move(model)),
+        vocab(std::move(vocab)),
+        params(params),
+        state(make_unique<llama_state>())
+    {
+    }
+    ~llama_context(){
+        ggml_free(model.ctx);
+    }
+};
+
+/* Original code by @ggerganov */
+// load the model's weights from a file
+bool llama_model_load(const std::string & fname, llama_model & model, gpt_vocab & vocab, int n_ctx) {
```

Comment:
I don't see latest changes on [master](https://github.com/ggerganov/llama.cpp/commit/0b366e735729327476ec31da02de3c9c9771ddfb) here, which means you did not update it properly.

Reply:
It's entirely possible I messed something up when merging in changes. The rate at which this repo is updating makes keeping up a Sisyphean task 😅. I will take a look.

Reply:
I have updated `llama_model_load()`. It was indeed missing a couple of features.